src/visit.rs

```diff
@@ -0,0 +1,211 @@
+use std::collections::BTreeMap;
+
+use crate::{FeatureValue, Metadata, Package};
+
+/// A visitor over a package's features and their dependencies.
+pub trait FeatureVisitor {
+    /// The error type of a walk.
+    type Error;
+
+    /// Visits a missing dependency.
+    ///
+    /// Return `Ok(())` to continue the walk, or `Err(…)` to abort it.
+    fn visit_missing_dependency(&mut self, dep_name: &str) -> Result<(), Self::Error>;
+
+    /// Visits a missing package.
+    ///
+    /// Return `Ok(())` to continue the walk, or `Err(…)` to abort it.
+    fn visit_missing_package(&mut self, pkg_name: &str) -> Result<(), Self::Error>;
```

Comment:
I'm confused about the terminology of `missing dependency` and `missing package`. In what sense are they missing, given that we seem to know their names based on the function signatures?

While all the trait items are documented, it would be good to frame their documentation in terms of "what does the user need to understand to choose a good implementation here." Right now the documentation seems to be written from the perspective of "what does the code *driving* the visitor need to know to drive it properly", which isn't that helpful for end-users.

Reply:
The documentation was mostly copied from the corresponding APIs on `cargo` itself, but the names used in the implementation slightly drifted afterwards. I've cleaned things up and also added more information. 👍 

The need for `visit_missing_dependency(…)` and `visit_missing_package(…)` is twofold:

1. The `Metadata` only contains `packages` for crates that were part of the active crate graph, i.e. crates that either were workspace members, non-optional dependencies or optional dependencies that where enabled via corresponding features. So if you walk the feature graph from a feature that would enable a dependency but that dependency wasn't also activated via `MetadataCommand.features(…)`, then lookup of the corresponding `Package` on `Metadata` will fail. But we need the `Package` to look up its features for the recursion. (I'd expect `visit_missing_dependency(…)` to be unnecessary for valid `Metadata` instances and haven't seen it fail myself so far.)

2. The `Metadata` might have been deserialized from an unchecked JSON blob with packages/dependencies missing. I'll leave deciding on whether or not we want to gracefully handle such situations to @oli-obk, but wanted to avoid pestering the draft implementation with `.unwrap()`s.

These failure modes are also partly what made me switch to a visitor API. 

I had initially implemented a single method on `Metadata` but the implementation very quickly became unwieldy and inflexible:

```rust
fn transitive_feature_dependencies<'a>(&'a self, package: &'a Package, feature_name: &str) -> Result<BTreeMap<&'a Package, BTreeSet<FeatureValue>>> {
    // ...
}
```

But whether or not this method should fail on missing packages depends on the use-case, I think.
With a visitor this decision is totally up to you, without bloating the crate's API with options.

Similarly one may want to be able to pick between different collection modes: recursive (i.e. transitive) and non-recursive (i.e. immediate) feature dependencies. Again, a visitor API leaves this up to the user, without bloating the crate's API with options.